File: _race_probe.js
/* Drives shading at several sizes so the injected race has something to
   measure, then reports what __RACE collected. */
(function () {
  var out = [], errs = [];
  var _ce = console.error;
  console.error = function () { errs.push(Array.prototype.join.call(arguments, ' ')); _ce.apply(console, arguments); };
  function log(k, v) { out.push(k + '=' + v); }
  var k, A, THREE;

  function freshCube() {
    var o = k.createCubeObject('R', new THREE.Vector3(0, 0, 0));
    A.selectedObjectIds = new Set([o.id]);
    A.activeObjectId = o.id;
    k.refreshUI();
    return o;
  }
  function subdivideOnce(o) {
    k.setMode('object');
    A.selectedObjectIds = new Set([o.id]);
    A.activeObjectId = o.id;
    A.selectedElements = new Set();
    k.subdivideSelection();
    if (!A.pendingOp) return false;
    k.flushPendingApply();
    k.confirmPendingOp();
    return true;
  }
  function tris(o) { return o.mesh.geometry.index.count / 3; }

  function main() {
    k = window.__kubik; A = k.App; THREE = k.THREE;

    log('0.injected', window.__RACE ? 'the legacy loop is in the page and running'
      : 'NO RACE OBJECT - nothing was injected, every number below is meaningless');

    var rows = [];
    var o = freshCube();
    for (var d = 0; d < 5; d++) {
      subdivideOnce(o);
      /* WARM FIRST, MEASURE SECOND. The first row of this table came out at
         -450% and it was the harness, not the code: it is the first thing
         the page measures, so the new loop was being JIT-compiled inside the
         window being timed while the old loop - which runs first in each
         shade - had already warmed every helper they share. _time_probe's
         own timeIt takes a MEDIAN of several runs for exactly this reason;
         this takes a mean, so it has to warm up explicitly. */
      for (var w = 0; w < 4; w++) k.applyShading(o);
      var before = window.__RACE ? { o: window.__RACE.oldMs, n: window.__RACE.newMs, r: window.__RACE.runs } : null;
      for (var i = 0; i < 8; i++) k.applyShading(o);
      if (before) {
        var dOld = window.__RACE.oldMs - before.o;
        var dNew = window.__RACE.newMs - before.n;
        var runs = window.__RACE.runs - before.r;
        rows.push(tris(o) + 'tri  old ' + (dOld / runs).toFixed(2) + 'ms  new ' +
          (dNew / runs).toFixed(2) + 'ms  ' +
          (dOld > 0 ? (100 * (dOld - dNew) / dOld).toFixed(0) + '% off' : '-'));
      }
      if (tris(o) > 20000) break;
    }
    log('1.race', rows.join('  |  '));

    if (window.__RACE) {
      var R = window.__RACE;
      log('2.totals', R.runs + ' shades: old ' + R.oldMs.toFixed(1) + 'ms, new ' +
        R.newMs.toFixed(1) + 'ms');
      log('2.faster', R.oldMs > 0
        ? (100 * (R.oldMs - R.newMs) / R.oldMs).toFixed(0) + '% off the union-find, measured in ONE run against the loop it replaces'
        : 'NO TIME RECORDED');
      /* THE HALF THAT MATTERS MORE. A faster answer that is a different
         answer is not an optimisation. Every normal component of every shade
         is compared; the worst absolute difference across all of them is
         what this reports. */
      log('3.identical', R.worst === 0
        ? 'every one of ' + R.checked + ' normal components came out BIT FOR BIT identical'
        : 'THE ANSWERS DIFFER - worst component off by ' + R.worst);
    }

    /* ---- 4. THE STAMP SURVIVES A THROW. The review's one real finding,
       and nothing else would have caught it: applyShading's catch
       deliberately lets the app carry on, so a throw part way through used
       to leave the shared stamp counter at the value the dead call STARTED
       from. The next shade re-issued those numbers, `enrol` found its own
       stamp already there and skipped the reset, and the union-find began on
       the dead call's forest - un-normalised normals from two calls, written
       with no error on screen because the catch had fired on the previous
       one.

       The throw is injected into the harness's own copy of the loop, so the
       shipped file carries no test hook. (The first attempt tried to reach
       geo.userData.shadeTopo from out here, which is null outside a direct
       drag - so it reported "no cached topo" and the two assertions below
       passed having tested nothing at all.) */
    var o4 = freshCube();
    subdivideOnce(o4); subdivideOnce(o4);
    k.applyShading(o4);
    var good = o4.mesh.geometry.attributes.normal.array.slice();

    var lever = (typeof window.__FORCE_THROW !== 'undefined') || true;
    window.__FORCE_THROW = 20;          // fail on the 20th logical vertex
    var errsBefore = errs.length;
    k.applyShading(o4);                 // applyShading swallows it by design
    var fired = window.__FORCE_THROW === 0;
    window.__FORCE_THROW = 0;
    log('4.forced_throw', fired
      ? 'a shade was made to fail on its 20th vertex, and applyShading swallowed it as it is designed to'
      : 'THE LEVER DID NOT FIRE - section 4 tested nothing (counter left at ' +
        window.__FORCE_THROW + ')');
    log('4.was_reported', errs.length > errsBefore
      ? 'and it was reported to the console, so the failure was real'
      : 'nothing reached the console');

    // A clean shade must now produce exactly what it did before the failure.
    k.applyShading(o4);
    var after = o4.mesh.geometry.attributes.normal.array;
    var worst4 = 0;
    for (var i4 = 0; i4 < good.length; i4++) {
      var d4 = Math.abs(after[i4] - good[i4]);
      if (d4 > worst4) worst4 = d4;
    }
    log('4.recovers', fired && worst4 === 0
      ? 'the shade after a failed one is bit-for-bit what it was before it - the stamp namespace did not collide'
      : (fired ? 'STALE STAMPS: the next shade differs by up to ' + worst4
        : 'not tested - the lever did not fire'));
    var worstLen = 0;
    for (var j4 = 0; j4 + 2 < after.length; j4 += 3) {
      var L4 = Math.sqrt(after[j4] * after[j4] + after[j4 + 1] * after[j4 + 1] + after[j4 + 2] * after[j4 + 2]);
      if (L4 > 1e-6) worstLen = Math.max(worstLen, Math.abs(L4 - 1));
    }
    log('4.unit_normals', fired && worstLen < 1e-5
      ? 'and every normal is still a unit vector - an un-normalised island is exactly what a collision produced'
      : (fired ? 'NORMALS ARE NOT UNIT: worst off by ' + worstLen : 'not tested'));
    void lever;

    log('console.errors', errs.length ? errs.join(' | ').slice(0, 300) : 'none');
  }

  var posted = false;
  function post() {
    if (posted) return;
    posted = true;
    try { fetch('/result', { method: 'POST', body: out.join('\n') }); }
    catch (e) { /* nothing else to try */ }
  }
  setTimeout(function () {
    if (!posted) { out.push('WATCHDOG=main did not finish - the last line above is where it hung'); post(); }
  }, 60000);
  function ready(cb, t) {
    t = t || 0;
    if (window.__kubik && window.__kubik.App) return cb();
    if (t > 400) { out.push('ERROR=no __kubik'); return post(); }
    setTimeout(function () { ready(cb, t + 1); }, 20);
  }
  window.addEventListener('error', function (e) { errs.push('onerror: ' + (e.message || e)); });
  setTimeout(function () {
    ready(function () {
      setTimeout(function () {
        try { main(); }
        catch (e) { out.push('ERROR=' + (e && e.stack ? e.stack.split('\n').slice(0, 8).join(' / ') : e)); }
        post();
      }, 800);
    });
  }, 400);
})();

File: _race_probe.py
# Races a2.82's rewritten union-find against the one it replaces, in ONE run
# on the SAME geometry, and compares the normals it produces vertex for
# vertex. index.html is never touched - the legacy loop is injected into a
# copy in memory, exactly as _prof_probe.py does its instrumentation.
#
# This is the method a2.77 used and the reason its 20% is believable: a
# cross-version timing on this machine drifts 3x, and "faster" is worthless
# without "and identical".
import io, os, shutil, tempfile, sys, subprocess, threading
import http.server, socketserver

ROOT = r'C:\Users\a.bodrov\Projects\kubik'
# A FRESH profile of this probe's own. Sharing Chrome's default profile
# makes the run exit in seconds whenever a Chrome is open; a FIXED
# directory makes the next run open this run's autosaved scene.
PROF = tempfile.mkdtemp(prefix='_prof_')
CHROME = r'C:\Program Files\Google\Chrome\Application\chrome.exe'
PORT = 8861
OUT = '_race_out.txt'

src = io.open(os.path.join(ROOT, 'index.html'), encoding='utf-8').read()

# NOTE: this anchor sits on the very line a2.82c changed (stamp++ became
# stamp = ++UF.stamp0), and the harness aborted loudly rather than injecting
# nothing - which is the behaviour to keep. If the union-find is edited
# again, expect ANCHOR MISS and update this first.
ANCHOR = """    trisAt.forEach((tris, l) => {
      stamp = ++UF.stamp0;"""

# A lever to make one shade fail PART WAY THROUGH the vertex loop, injected
# into the copy so the shipped file carries no test hook. The first attempt
# reached for geo.userData.shadeTopo from the probe instead - which is null
# outside a direct drag (a2.77), so the section passed while testing nothing.
THROW_LEVER = """      if (window.__FORCE_THROW && --window.__FORCE_THROW === 0) {
        throw new Error('forced mid-shade failure');
      }
"""

LEGACY = """    /* ---- INJECTED BY _race_probe.py - not in the shipped file ---- */
    window.__RACE = window.__RACE || { oldMs: 0, newMs: 0, runs: 0, worst: 0, checked: 0 };
    const _baseline = normals.slice();
    const _r0 = performance.now();
    {
      // The loop a2.82 replaces, reading the SAME records so the comparison
      // isolates the data structures rather than the two changes together.
      const parent = new Map();
      const findOld = (g) => { while (parent.get(g) !== g) { parent.set(g, parent.get(parent.get(g))); g = parent.get(g); } return g; };
      const acc = new Map();
      trisAt.forEach((tris, l) => {
        parent.clear(); acc.clear();
        for (let i = 0; i < tris.length; i++) { const g = triGroup[tris[i]]; if (!parent.has(g)) parent.set(g, g); }
        const inc = incident.get(l);
        if (inc) for (let i = 0; i < inc.length; i++) {
          const rec = inc[i];
          if (rec.sharp) continue;
          const fs = rec.faces;
          for (let j = 0; j < fs.length; j++) if (!parent.has(fs[j])) parent.set(fs[j], fs[j]);
          for (let j = 1; j < fs.length; j++) {
            const r1 = findOld(fs[0]), r2 = findOld(fs[j]);
            if (r1 !== r2) parent.set(r1, r2);
          }
        }
        for (let i = 0; i < tris.length; i++) {
          const t = tris[i], r = findOld(triGroup[t]);
          let v = acc.get(r);
          if (!v) { v = new THREE.Vector3(); acc.set(r, v); }
          v.add(triNormal[t]);
        }
        acc.forEach(v => { if (v.lengthSq() > 1e-12) v.normalize(); });
        const ais = logicalGroups[l];
        for (let i = 0; i < ais.length; i++) {
          const ai = ais[i], g = groupOfAttr[ai];
          if (g < 0 || !parent.has(g)) continue;
          const v = acc.get(findOld(g));
          if (!v || v.lengthSq() < 1e-12) continue;
          normals[ai * 3] = v.x; normals[ai * 3 + 1] = v.y; normals[ai * 3 + 2] = v.z;
        }
      });
    }
    const _r1 = performance.now();
    const _oldOut = normals.slice();
    normals.set(_baseline);          // back to what the new loop will start from
    const _r2 = performance.now();
    /* ---- and now the shipped loop, timed from here ---- */
"""

TAIL = """    {
      const _r3 = performance.now();
      window.__RACE.oldMs += (_r1 - _r0);
      window.__RACE.newMs += (_r3 - _r2);
      window.__RACE.runs++;
      let worst = 0;
      for (let i = 0; i < normals.length; i++) {
        const d = Math.abs(normals[i] - _oldOut[i]);
        if (d > worst) worst = d;
      }
      window.__RACE.worst = Math.max(window.__RACE.worst, worst);
      window.__RACE.checked += normals.length;
    }
"""

if src.count(ANCHOR) != 1:
    print('ANCHOR MISS (%d)' % src.count(ANCHOR)); sys.exit(1)
src = src.replace(ANCHOR, LEGACY + ANCHOR + '\n' + THROW_LEVER, 1)

END = """    geo.attributes.normal.needsUpdate = true;"""
if src.count(END) != 1:
    print('END MISS (%d)' % src.count(END)); sys.exit(1)
src = src.replace(END, TAIL + END, 1)
print('injected the legacy loop')

JS = io.open(os.path.join(ROOT, '_race_probe.js'), encoding='utf-8').read()
assert '</body>' in src
io.open(os.path.join(ROOT, '_race_probe.html'), 'w', encoding='utf-8', newline='').write(
    src.replace('</body>', '<script>\n' + JS + '\n</script>\n</body>', 1))

result = {'text': None}
done = threading.Event()

class H(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *a, **kw):
        super().__init__(*a, directory=ROOT, **kw)
    def log_message(self, *a):
        pass
    def do_POST(self):
        n = int(self.headers.get('Content-Length', 0))
        result['text'] = self.rfile.read(n).decode('utf-8', 'replace')
        self.send_response(204)
        self.end_headers()
        done.set()

socketserver.TCPServer.allow_reuse_address = True
srv = socketserver.TCPServer(('127.0.0.1', PORT), H)
threading.Thread(target=srv.serve_forever, daemon=True).start()

url = 'http://127.0.0.1:%d/_race_probe.html?debug=1' % PORT
cmd = [CHROME, '--headless=new', '--disable-gpu-sandbox', '--no-sandbox',
       '--enable-unsafe-swiftshader', '--use-angle=swiftshader',
       '--window-size=900,700', '--user-data-dir=' + PROF,
       # SHARED, and deliberately not the fresh profile: the app pulls three.js
       # from a CDN on every load, and a cold cache is what actually broke
       # _imp_probe. State fresh, network warm.
       '--disk-cache-dir=' + os.path.join(ROOT, '_httpcache'),
       '--no-first-run', '--no-default-browser-check', '--disable-sync',
       '--disable-background-networking', '--disable-component-update',
       '--disable-default-apps', '--disable-extensions', '--metrics-recording-only',
       '--mute-audio',
       url]
p = subprocess.Popen(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
ok = done.wait(timeout=180)
try:
    p.terminate(); p.wait(timeout=20)
except Exception:
    p.kill()
srv.shutdown()
shutil.rmtree(PROF, ignore_errors=True)

txt = result['text'] if ok and result['text'] else 'NO PROBE OUTPUT (timed out)'
io.open(os.path.join(ROOT, OUT), 'w', encoding='utf-8').write(txt)
print(txt)
